### PR TITLE
Implement fluentd logging driver

### PR DESCRIPTION
**Pull request by Zheaoli**
Related to #1039 

Need to be merged after #1062

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Thanks, but let me mark this as a draft until we review/merge https://github.com/containerd/nerdctl/pull/1062 (sorry for delay in reviewing/merging 🙇 )

**Comment by Zheaoli** [comment]
> Thanks, but let me mark this as a draft until we review/merge #1062 (sorry for delay in reviewing/merging 🙇 )

No problem! Thanks for your work!

File: README.md [hunk 1]
```diff
@@ -473,13 +473,22 @@ Metadata flags:
...
+        - :whale: `--log-opt=max-file=<MAX-FILE>`: The maximum number of log files that can be present. If rolling the logs creates excess files, the oldest file is removed. Only effective when `max-size` is also set. A positive integer. Defaults to 1.
     - :whale: `--log-driver=journald`: Writes log messages to `journald`. The `journald` daemon must be running on the host machine.
-          - :whale: `--log-opt=tag=<TEMPLATE>`: Specify template to set `SYSLOG_IDENTIFIER` value in journald logs.
+      - :whale: `--log-opt=tag=<TEMPLATE>`: Specify template to set `SYSLOG_IDENTIFIER` value in journald logs.
+    - :whale: `--log-driver=fluentd`: Writes log messages to `fluentd`. The `fluentd` daemon must be running on the host machine.
+      - The `fluentd` logging driver supports the following logging options:
+        - :whale: `--log-opt=fluentd-address=<ADDRESS>`: The address of the `fluentd` daemon.
+        - :whale: `--log-opt=fluentd-async=<true|false>`: Enable async mode for fluentd. The default value is false.
+        - :whale: `--fluentd-retry-wait=<1s|1ms>`: The time to wait before retrying to send logs to fluentd. The default value is 1s.
```
Comment (AkihiroSuda) [inline README.md:487]: Did you mean:
```suggestion
        - :whale: `--log-opt=fluentd-retry-wait=<1s|1ms>`: The time to wait before retrying to send logs to fluentd. The default value is 1s.
```

File: README.md [hunk 2]
```diff
@@ -473,13 +473,22 @@ Metadata flags:
...
+        - :whale: `--fluentd-max-retries=<1>`: The maximum number of retries to send logs to fluentd. The default value is MaxInt32.
+        - :whale: `--fluentd-sub-second-precision=<true|false>`: Enable sub-second precision for fluentd. The default value is false.
+        - :whale: `--log-opt=fluentd-async-reconnect-interval=<1s|1ms>`: The time to wait before retrying to reconnect to fluentd. The default value is 0s.
```
Comment (AkihiroSuda) [inline README.md:490]: Is this really `:whale` or `:nerd_face:` extension?
Doesn't seem documented in https://docs.docker.com/config/containers/logging/fluentd/

File: README.md [hunk 3]
```diff
@@ -473,13 +473,22 @@ Metadata flags:
...
+        - :whale: `--log-opt=fluentd-request-ack=<true|false>`: Enable request ack for fluentd. The default value is false.
```
Comment (AkihiroSuda) [inline README.md:491]: same as above

File: go.mod [hunk 4]
```diff
@@ -137,7 +138,8 @@ require (
 	github.com/opencontainers/runtime-tools v0.0.0-20190417131837-cd1349b7c47e // indirect
 	github.com/opencontainers/selinux v1.10.1 // indirect
 	github.com/opentracing/opentracing-go v1.2.0 // indirect
-	github.com/pkg/errors v0.9.1 // indirect
+	github.com/philhofer/fwd v1.1.1 // indirect
+	github.com/pkg/errors v0.9.1
```
Comment (AkihiroSuda) [inline go.mod:142]: deprecated and archived

File: pkg/testutil/testutil.go [hunk 5]
```diff
@@ -37,6 +37,8 @@ import (
 	"gotest.tools/v3/icmd"
 )
 
+var FluentdImage = "fluent/fluentd:v1.14-1"
```
Comment (AkihiroSuda) [inline pkg/testutil/testutil.go:40]: This should be in `testutil_linux.go`

Comment (AkihiroSuda) [reply pkg/testutil/testutil.go:40]: ping @ktock for mirroring

Comment (ktock) [reply pkg/testutil/testutil.go:40]: Mirrored: [`ghcr.io/stargz-containers/fluent/fluentd:v1.14-1-org`](https://github.com/orgs/stargz-containers/packages/container/package/fluent%2Ffluentd)
 cc @Zheaoli

Comment (Zheaoli) [reply pkg/testutil/testutil.go:40]: Got it

File: README.md (hunk 1 above)
Comment (Zheaoli) [reply README.md]: My bad, typo, dome

File: README.md (hunk 2 above)
Comment (Zheaoli) [reply README.md]: fixed

File: README.md (hunk 3 above)
Comment (Zheaoli) [reply README.md]: fixed

File: pkg/logging/fluentd_logger.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+		scanner := bufio.NewScanner(reader)
+		for scanner.Scan() {
+			if scanner.Err() != nil {
+				return
+			}
+			metaData["log"] = scanner.Text()
+			fluentClient.PostWithTime(f.Opts[Tag], time.Now(), metaData)
+		}
+	}
+	go fun(&wg, config.Stdout, config, "stdout")
```
Comment (fahedouch) [inline pkg/logging/fluentd_logger.go:102]: `go fun(&wg, config, "stdout")` : `config` contains the Readers no need to pass them as parameters

File: pkg/logging/fluentd_logger.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+	port     int
+	path     string
+}
+
+func (f *FluentdLogger) Process(_ string, config *logging.Config) error {
+	if runtime.GOOS == "windows" {
+		// TODO: support fluentd on windows
+		return fmt.Errorf("logging to fluentd is not supported on windows")
+	}
+	fluentConfig, err := parseFluentdConfig(f.Opts)
```
Comment (fahedouch) [inline pkg/logging/fluentd_logger.go:76]: is it possible to check if fluentd  daemon is `Enabled` before starting process ?

File: README.md [hunk 8]
```diff
@@ -473,13 +473,23 @@ Metadata flags:
...
+        - :whale: `--log-opt=max-size=<MAX-SIZE>`: The maximum size of the log before it is rolled. A positive integer plus a modifier representing the unit of measure (k, m, or g). Defaults to unlimited.
+        - :whale: `--log-opt=max-file=<MAX-FILE>`: The maximum number of log files that can be present. If rolling the logs creates excess files, the oldest file is removed. Only effective when `max-size` is also set. A positive integer. Defaults to 1.
     - :whale: `--log-driver=journald`: Writes log messages to `journald`. The `journald` daemon must be running on the host machine.
-          - :whale: `--log-opt=tag=<TEMPLATE>`: Specify template to set `SYSLOG_IDENTIFIER` value in journald logs.
+      - :whale: `--log-opt=tag=<TEMPLATE>`: Specify template to set `SYSLOG_IDENTIFIER` value in journald logs.
+    - :whale: `--log-driver=fluentd`: Writes log messages to `fluentd`. The `fluentd` daemon must be running on the host machine.
```
Comment (fahedouch) [inline README.md:483]: can we mention the input plugins that we support, obviously, `tcp` and `unix`

File: pkg/logging/fluentd_logger.go [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+		}
+	}
+	return nil
+}
+
+func parseFluentdConfig(config map[string]string) (fluent.Config, error) {
+	result := fluent.Config{}
+	location, err := parseAddress(config[fluentAddress])
+	if err != nil {
+		return result, errors.Wrapf(err, "invalid fluentd address (%s)", config[fluentAddress])
```
Comment (fahedouch) [inline pkg/logging/fluentd_logger.go:184]: same as above for error formatting

File: pkg/logging/fluentd_logger.go (hunk 7 above)
Comment (Zheaoli) [reply pkg/logging/fluentd_logger.go:76]: I think it's a little bit hard.

File: README.md (hunk 8 above)
Comment (Zheaoli) [reply README.md:483]: Yes, we can

File: pkg/logging/fluentd_logger.go [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+		// TODO: support fluentd on windows
+		return fmt.Errorf("logging to fluentd is not supported on windows")
+	}
+	fluentConfig, err := parseFluentdConfig(f.Opts)
+	if err != nil {
+		return err
+	}
+	fluentClient, err := fluent.New(fluentConfig)
+	if err != nil {
+		return fmt.Errorf("create fluent client failed, error is %v", err)
```
Comment (fahedouch) [inline pkg/logging/fluentd_logger.go:81]: `return fmt.Errorf("failed to  create fluent client: %w", err)`

**Review by fahedouch — APPROVED** [review]

**Comment by fahedouch** [comment]
Thanks

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**